scipy/stats/_distn_infrastructure.py

```diff
@@ -2386,18 +2445,48 @@ def fit(self, data, *args, **kwds):
         loc = kwds.pop('loc', start[-2])
         scale = kwds.pop('scale', start[-1])
         args += (loc, scale)
-        x0, func, restore, args = self._reduce_func(args, kwds)
+        x0, func, restore, args = self._reduce_func(args, kwds, data=data)
         optimizer = kwds.pop('optimizer', optimize.fmin)
         # convert string to function in scipy.optimize
         optimizer = _fit_determine_optimizer(optimizer)
         # by now kwds must be empty, since everybody took what they needed
         if kwds:
             raise TypeError("Unknown arguments: %s." % kwds)
 
-        vals = optimizer(func, x0, args=(ravel(data),), disp=0)
+        # method of moments could be done with fsolve/root instead of an
+        # optimizer, but using an optimizer reduces maintains the structure of
+        # MM/MLE and paves the way for Generalized Method of Moments
+        vals, obj = optimizer(func, x0, args=(ravel(data),), disp=0,
+                              full_output=True)[:2]
+
         if restore is not None:
             vals = restore(args, vals)
         vals = tuple(vals)
+
+        a, b = self.support(*vals)
+        beyond_support = np.count_nonzero((data < a) | (data > b))
+        if beyond_support:
+            raise Exception(f"Optimization converged to parameters that are "
+                            f"inconsistent with the data: {beyond_support} "
+                            f"data points are outside the support of the "
+                            f"fitted distribution.")
+
+        loc, scale, shapes = self._unpack_loc_scale(vals)
+        if not (np.all(self._argcheck(*shapes)) and scale > 0):
+            raise Exception("Optimization converged to parameters that are "
+                            "invalid for the distribution.")
+
+        if method == 'mm':
+            if not np.isfinite(obj):
+                raise Exception("Optimization failed: either a data moment "
```

Comment:
I think the data moments will always be finite (non-nan, even), actually.

Reply:
There is always the possibility that the calculation of the data moment could overflow, giving `inf`.  And if two consecutive terms in the computation of an odd moment result in `[inf, -inf]`, the final result will be `nan`.  E.g.

```
In [36]: x = np.full(10, fill_value=1e160)

In [37]: x[:5] = -x[:5] 

In [38]: x
Out[38]: 
array([-1.e+160, -1.e+160, -1.e+160, -1.e+160, -1.e+160,  1.e+160,
        1.e+160,  1.e+160,  1.e+160,  1.e+160])

In [39]: [np.mean(x**i) for i in range(1, 5)]  # Raw sample moments 1, 2, 3, 4.    
/Users/warren/mc37allpip/bin/ipython:1: RuntimeWarning: overflow encountered in square
  #!/Users/warren/mc37allpip/bin/python
/Users/warren/mc37allpip/bin/ipython:1: RuntimeWarning: overflow encountered in power
  #!/Users/warren/mc37allpip/bin/python
/Users/warren/mc37allpip/lib/python3.7/site-packages/numpy-1.20.0.dev0+c22fab7-py3.7-macosx-10.9-x86_64.egg/numpy/core/_methods.py:160: RuntimeWarning: invalid value encountered in reduce
  ret = umr_sum(arr, axis, dtype, out, keepdims)
Out[39]: [0.0, inf, nan, inf]
```